runtime/pyproject.toml

```diff
@@ -3,9 +3,10 @@ requires = [
     "setuptools>=42",
     "wheel",
     "cmake",
+    "nanobind>=1.4.0",
     "ninja",
+    "numpy",
     "packaging",
-    "pybind11>=2.10.1",
```

Comment:
Can you add a brief rationale section to the PR description and link to https://github.com/pybind/pybind11 and https://github.com/wjakob/nanobind (or similar references)? 

nanobind's README is nearly what I'm looking for, though benchmark/profile results specific to IREE would be nice if you have them:
> _nanobind_ is a small binding library that exposes C++ types in Python and vice versa. It is reminiscent of [Boost.Python](https://www.boost.org/doc/libs/1_64_0/libs/python/doc/html) and [pybind11](https://github.com/pybind/pybind11) and uses near-identical syntax. In contrast to these existing tools, nanobind is more efficient: bindings compile in a shorter amount of time, produce smaller binaries, and have better runtime performance.